src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverter.cs

```diff
@@ -40,6 +40,73 @@ public override void Write(Utf8JsonWriter writer, object? value, JsonSerializerO
             writer.WriteEndObject();
         }
 
+        public sealed override void WriteAsPropertyName(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        {
+            WriteAsPropertyNameCore(writer, value, options, isWritingExtensionDataProperty: false);
+        }
+
+        internal sealed override void WriteAsPropertyNameCore(Utf8JsonWriter writer, object value, JsonSerializerOptions options, bool isWritingExtensionDataProperty)
+        {
+            if (value is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(value));
+            }
+
+            Type runtimeType = value.GetType();
+            if (runtimeType == TypeToConvert)
+            {
+                ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(runtimeType, this);
+            }
+
+            JsonConverter runtimeConverter = options.GetConverterInternal(runtimeType);
+            runtimeConverter.WriteAsPropertyNameCoreAsObject(writer, value, options, isWritingExtensionDataProperty);
+        }
+    }
+
+    /// <summary>
+    /// Defines an object converter that only supports (polymorphic) serialization but not deserialization.
+    /// This is done to avoid rooting dependencies to JsonNode/JsonElement necessary to drive object deserialization.
+    /// Source generator users need to explicitly declare support for object so that the derived converter gets used.
+    /// </summary>
+    internal sealed class SlimObjectConverter : ObjectConverter
+    {
+        // Keep track of the originating resolver so that the converter surfaces
+        // an accurate error message whenever deserialization is attempted.
+        private readonly IJsonTypeInfoResolver _originatingResolver;
+
+        public SlimObjectConverter(IJsonTypeInfoResolver originatingResolver)
+            => _originatingResolver = originatingResolver;
+
+        public override object? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            ThrowHelper.ThrowNotSupportedException_NoMetadataForType(typeToConvert, _originatingResolver);
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Defines an object converter that supports deserialization via JsonElement/JsonNode representations.
+    /// Used as the default in reflection or if object is declared in the JsonSerializerContext type graph.
+    /// </summary>
+    internal sealed class DeserializingObjectConverter : ObjectConverter
```

Comment:
(super nit) Does "DeserializingObjectConverter" make it sound like it only supports deserializing and not serialization? Since this is opposed to "SlimObjectConverter", what do you think about "FullObjectConverter"? Or "DefaultObjectConverter"?

Reply:
DefaultObjectConverter sounds better, changed to that.